iothub/service/src/Http/HttpMessageHelper.cs

```diff
@@ -80,48 +75,30 @@ internal static async Task<T> DeserializeResponseAsync<T>(HttpResponseMessage re
         /// </param>
         /// <exception cref="ArgumentNullException">Thrown when the provided <paramref name="requestMessage"/> or <paramref name="requestMessage"/> is null.</exception>
         /// <exception cref="ArgumentException">Thrown if the <paramref name="eTag"/> is empty or white space.</exception>
         internal static void ConditionallyInsertETag(HttpRequestMessage requestMessage, ETag eTag, bool onlyIfUnchanged)
         {
             Debug.Assert(requestMessage != null, "Request message should not have been null");
 
-            if (onlyIfUnchanged && !string.IsNullOrWhiteSpace(eTag.ToString()))
+            if (!onlyIfUnchanged)
             {
-                // "Perform this operation only if the entity is unchanged"
-                // Sends the If-Match header with a value of the ETag.
-                string escapedETag = EscapeETag(eTag.ToString());
-                requestMessage.Headers.IfMatch.Add(new EntityTagHeaderValue(escapedETag, true));
+                eTag = ETag.All;
             }
-            else
-            {
-                // "Perform this operation even if the entity has changed"
-                // Sends the If-Match header with a value of "*"
-                requestMessage.Headers.IfMatch.Add(new EntityTagHeaderValue(s_eTagForce, true));
-            }
-        }
 
-        // ETag values other than "*" need to be wrapped in escaped quotes if they are not
-        // already.
-        private static string EscapeETag(string eTag)
-        {
-            var escapedETagBuilder = new StringBuilder();
-
-            if (!eTag.StartsWith("\"", StringComparison.OrdinalIgnoreCase))
-            {
-                escapedETagBuilder.Append('"');
-            }
-
-            escapedETagBuilder.Append(eTag);
-
-            if (!eTag.EndsWith("\"", StringComparison.OrdinalIgnoreCase))
+            if (!string.IsNullOrWhiteSpace(eTag.ToString()))
             {
-                escapedETagBuilder.Append('"');
+                // Azure.Core.ETag expects the format "H" for serializing ETags that go into the header.
+                // https://github.com/Azure/azure-sdk-for-net/blob/9c6238e0f0dd403d6583b56ec7902c77c64a2e37/sdk/core/Azure.Core/src/ETag.cs#L87-L114
+                // Also, System.Net.Http.Headers does not allow ETag.All (*) as a valid value even though RFC allows it.
+                // For this reason, we'll add the ETag value without additional validation.
+                // System.Net.Http.Headers validation: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Http/tests/UnitTests/Headers/EntityTagHeaderValueTest.cs#L214,
+                // https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Http/src/System/Net/Http/Headers/GenericHeaderParser.cs#L98
+                // RFC specification: https://www.rfc-editor.org/rfc/rfc7232#section-3.1
+                _ = requestMessage.Headers.TryAddWithoutValidation("If-Match", eTag.ToString("H"));
```

Comment:
Should we throw an exception if this "TryAdd" fails? Previously this threw IEnumerable.Add exceptions or an exception that HttpRequestHeaders defined, eg. FormatException.

Reply:
That's a tough one. Should a user know if we can't put their eTag in and it just succeeds anyway? Maybe so.

Will that be a scenario blocker though? Which is worse?

Reply:
This adds the etag without any validation. So the only scenario I'd except this to fail over is if we try to add the etag twice - key already exists scenario. Since our SDK doesn't do that, I'm ok with us not inspecting the result of this call.